25-class-metaprog/persistent/dblib.py

```diff
@@ -0,0 +1,151 @@
+# SQLite3 does not support parameterized table and field names,
+# for CREATE TABLE and PRAGMA so we must use Python string formatting.
+# Applying `check_identifier` to parameters prevents SQL injection.
+
+import sqlite3
+from typing import NamedTuple
+
+DEFAULT_DB_PATH = ':memory:'
+CONNECTION = None
+
+SQL_TYPES = {
+    int: 'INTEGER',
+    str: 'TEXT',
+    float: 'REAL',
+    bytes: 'BLOB',
+}
+
+
+class NoConnection(Exception):
+    """Call connect() to open connection."""
+
+
+class SchemaMismatch(ValueError):
+    """The table's schema doesn't match the class."""
+
+    def __init__(self, table_name):
+        self.table_name = table_name
+
+
+class NoSuchRecord(LookupError):
+    """The given primary key does not exist."""
+
+    def __init__(self, pk):
+        self.pk = pk
+
+
+class UnexpectedMultipleResults(Exception):
+    """Query returned more than 1 row."""
+
+
+class ColumnSchema(NamedTuple):
+    name: str
+    sql_type: str
+
+
+def check_identifier(name):
+    if not name.isidentifier():
+        raise ValueError(f'{name!r} is not an identifier')
+
+
+def connect(db_path=DEFAULT_DB_PATH):
+    global CONNECTION
+    CONNECTION = sqlite3.connect(db_path)
+    return CONNECTION
+
+
+def get_connection():
+    if CONNECTION is None:
+        raise NoConnection()
+    return CONNECTION
+
+
+def gen_columns_sql(fields):
+    for name, py_type in fields.items():
+        check_identifier(name)
+        try:
+            sql_type = SQL_TYPES[py_type]
+        except KeyError as e:
+            raise ValueError(f'type {py_type!r} is not supported') from e
+        yield ColumnSchema(name, sql_type)
+
+
+def make_schema_sql(table_name, fields):
+    check_identifier(table_name)
+    pk = 'pk INTEGER PRIMARY KEY,'
+    spcs = ' ' * 4
+    columns = ',\n    '.join(
+        f'{field_name} {sql_type}'
+        for field_name, sql_type in gen_columns_sql(fields)
+    )
+    return f'CREATE TABLE {table_name} (\n{spcs}{pk}\n{spcs}{columns}\n)'
+
+
+def create_table(table_name, fields):
+    con = get_connection()
+    con.execute(make_schema_sql(table_name, fields))
+
+
+def read_columns_sql(table_name):
+    con = get_connection()
+    check_identifier(table_name)
+    rows = con.execute(f'PRAGMA table_info({table_name!r})')
+    # row fields: cid name type notnull dflt_value pk
+    return [ColumnSchema(r[1], r[2]) for r in rows]
```

Comment:
Alternative:

```suggestion
    return [ColumnSchema(name, type) for _, name, type, *_ in rows]
```

Reply:
I agree your code is more readable. sqlite3 also provides a row factory and a ready-to-use Row class that may work in this case. I will investigate.